### PR TITLE
Dynamic Dashboard: Dashboard view in SwiftUI

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #12411 

## Description

This PR migrates part of the dashboard view controller into a new SwiftUI view in preparation for the dynamic dashboard project. The existing `DashboardViewController` is massive and messy so hopefully this refactoring makes updating the UI much more easily.

Due to the large size of changes, this first PR handles half of the logic on dashboard view:
- Added a new enum for dashboard cards.
- Created a SwiftUI view and hosting controller.
- Switched to the new SwiftUI view for the dashboard screen with a feature flag.
- Added the following features to the new SwiftUI view:
  - Store name subtitle: the old view hides the subtitle when the title is small. I keep the subtitle in this new screen for simplicity.
  - Tab bar item on the new view.
  - Share store button and functionality.
  - Added store stats and top performers section by wrapping the existing view controller as-is.
  - Added store onboarding card.
  - Added Blaze section.
  - Handled pull-to-refresh.

The rest of the features will be added in a subsequent PR.

## Testing instructions

- Ensure that the feature flag `dynamicDashboard` is enabled.
- Build the app and log in to a store.
- Confirm that the dashboard screen contains a correct title, subtitle, functioning share button, with 3 sections: store stats & top performers, onboarding card and Blaze section if available.
- Confirm that the pull-to-refresh feature works as before.

Disable the feature flag `dynamicDashboard` and build the app. Confirm that the dashboard screen still works as before.

## Screenshots

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingView.swift [hunk 1]
```diff
@@ -183,7 +187,7 @@ private extension StoreOnboardingView {
 private extension StoreOnboardingView {
     enum Layout {
         static let insetsWithViewAllButton: EdgeInsets = .init(top: 16, leading: 16, bottom: 16, trailing: 16)
-        static let insetsWithoutViewAllButton: EdgeInsets = .init(top: 16, leading: 16, bottom: 0, trailing: 16)
+        static let insetsWithoutViewAllButton: EdgeInsets = .init(top: 16, leading: 16, bottom: 1, trailing: 16)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/Onboarding/StoreOnboardingView.swift:190]: The 0 padding causes an issue with the layout of the view when embedded in the dashboard view. Setting a non-zero layout is a workaround to fix it.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+                    ShareLink(item: url) {
+                        Image(systemName: "square.and.arrow.up")
+                    }
+                }
+            }
+        }
+        .onReceive(ServiceLocator.stores.site) { currentSite in
+            self.currentSite = currentSite
+        }
+        .refreshable {
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift:61]: TIL, this is convenient!

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift [hunk 3]
```diff
@@ -389,6 +392,13 @@ private extension StoreStatsAndTopPerformersViewController {
         view.backgroundColor = Constants.backgroundColor
     }
 
+    /// Resets the size of the view controller to display all contents on the `DashboardView`.
+    ///
+    func resetViewSize() {
+        let size = view.systemLayoutSizeFitting(UIView.layoutFittingExpandedSize)
+        preferredContentSize = size
+    }
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:400]: ❓  I don't fully understand this part of the code. Why does the view size need to be reset after sync?

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ViewControllerContainer.swift [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import SwiftUI
+
+/// A simple SwiftUI wrapper for any view controller
+struct ViewControllerContainer: UIViewControllerRepresentable {
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ViewControllerContainer.swift:4]: Interesting that we didn't have something generic like this before. 👍🏼

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        return NSLocalizedString("My Store", comment: "Title of the 'My Store' tab - used for spotlight indexing on iOS.")
+    }
+
+    var activityDescription: String? {
+        return NSLocalizedString("See at a glance which products are winning.",
+                                 comment: "Description of the 'My Store' screen - used for spotlight indexing on iOS.")
+    }
+
+    var activityKeywords: Set<String>? {
+        let keyWordString = NSLocalizedString("woocommerce, my store, today, this week, this month, this year," +
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift:152]: 💡 not related to this PR but I wonder if we should optionally add blaze and onboarding here when applicable.

**Review by hafizrahman — APPROVED** [review]
First of all, nice work on this. I followed it commit by commit and can imagine it's a decent amount of work migrating things to SwiftUI.

I like that there's a minimal amount of change to `DashboardViewModel` as well.

I worry a little that `DashboardViewHostingController` can be too big but so far it seems manageable.

I also tested the app and they work nicely. Individual functionalities inside each card is also replicated nicely.

I left some comments and a question but it's non-blocking.

Finally, also unrelated to this PR:
 
While testing I saw that `InAppFeedbackCardViewController` is added inside `StoreStatsAndTopPerformersPeriodViewController`. What do you think about separating that feedback card too and making it its own card?

File: WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/Stats v4/StoreStatsAndTopPerformersViewController.swift:400]: The reason is that after syncing, the section may have from 0-5 top performers, so the view height can vary. Resetting the view size makes sure that all contents can be displayed properly.

File: WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ViewControllerContainer.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/ReusableViews/SwiftUI Components/ViewControllerContainer.swift:4]: Previously we had specific view controllers that require custom input for their initializers, so we've been building separate `UIViewControllerRepresentable`-conforming views. This generic view is used for view controllers that don't need any input.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift (hunk 5 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewHostingController.swift:152]: From my understanding, these keywords are used in iPhone spotlight to quickly open the Woo app. If we want to make Blaze searchable, we should update the Blaze campaign list view screen instead. I'm not sure about how useful searching for store onboarding is.

**Comment by itsmeichigo** [comment]
Thank you Hafiz for reviewing the PR!

> While testing I saw that InAppFeedbackCardViewController is added inside StoreStatsAndTopPerformersPeriodViewController. What do you think about separating that feedback card too and making it its own card?

I'm keeping the store stats and top performers, including the feedback card together as-is for now to focus on refactoring the dashboard view controller only. Depending on Joe's design decision, we'll have to separate the sections and update the cards later for the dynamic dashboard project.

I'm merging this PR now to unblock development.

**Pull request merged**